wasi-nn.witx

```diff
@@ -58,6 +63,63 @@
     (param $target $execution_target)
     (result $error (expected $graph (error $nn_errno)))
   )
+  ;;; Load an opaque sequence of bytes to use for inference.
+  ;;;
+  ;;; This allows runtime implementations to support multiple graph encoding formats. For unsupported graph encodings,
+  ;;; return `errno::inval`.
+  (@interface func (export "load_by_name")
+     ;;; The name of the model to load from the model registry
+     (param $model_name $utf8name)
+
+     (result $error (expected $graph (error $nn_errno)))
+  )
+
+  ;;; Allows registering a model via URI
+  ;;;
+  (@interface func (export "register_model_uri")
+    ;;; The URI from where the model can be retrieved.
+    (param $uri $utf8uri)
+    (param $model_name $utf8name)
+    ;;; The encoding of the graph.
+    (param $encoding $graph_encoding)
+    ;;; Where to execute the graph.
+    (param $target $execution_target)
+
+    (result $error (expected (error $nn_errno)))
+  )
+
+  ;;; Allows registering a model directly via bytes.
+  ;;;
+  (@interface func (export "register_model_bytes")
+    ;;; The bytes necessary to build the graph.
+    (param $model_name $utf8name)
+    (param $builder $graph_builder_array)
+    ;;; The encoding of the graph.
+    (param $encoding $graph_encoding)
+    ;;; Where to execute the graph.
+    (param $target $execution_target)
+
+    (result $error (expected (error $nn_errno)))
+  )
+
+  ;;; Remove a model from the registry.
+  ;;;
+  (@interface func (export "unregister")
+    ;;; The bytes necessary to build the graph.
+    (param $model_name $utf8name)
+
+    (result $error (expected (error $nn_errno)))
+  )
+
+  ;;; Check if a model is registered with the registry.
+  ;;;
+  (@interface func (export "is_registered")
+    ;;; The bytes necessary to build the graph.
+    (param $model_name $utf8name)
+
+    (result $error (expected $status (error $nn_errno)))
+  )
```

Comment:
I would propose we leave `unregister` and `is_registered` out for the time being. Why? It's more work for people to implement and more things to change if we need to in the future. If names are mainly registered externally (outside of the Wasm context) then `unregister` will rarely if ever be used (which raises the question if we even need the function to register a model from within the Wasm context!). And `load_by_name` encompasses the functionality in `is_registered`: if you get a `model` back then, yes, the model was registered.

Reply:
The reason for these functions is to enable clients to pull models from an origin efficiently. I also made an update to add a u32 TTL to further reflect this.

Reply:
Correct. The wasi-nn spec wouldn't be responsible for retrieving bytes. Only for letting the host know that these bytes should be kept around as efficiently as possible in case they need re-use. The reason to support register / deregister / is_registered is loading a model is 3x - 5x as expensive as performing a single inference. One way to think about this is a way of letting the host know that this model may get re-used frequently. 

I would actually expect this to be the more common path than a one time load(...).

Reply:
Seems like the discussion is moving to https://github.com/WebAssembly/wasi-nn/issues/36... we can continue there.

Reply:
I have removed register/unregister for now. This can be handled via host configuration.